pgxn/neon/communicator.c

```diff
@@ -2499,16 +2548,29 @@ communicator_read_slru_segment(SlruKind kind, int64 segno, neon_request_lsns *re
 void
 communicator_reconfigure_timeout_if_needed(void)
 {
-	bool	needs_set = MyPState->ring_receive != MyPState->ring_unused &&
+	bool	needs_set = (MyPState->ring_receive != MyPState->ring_unused ||
+						 (MIN_BACKEND_PREFETCH_LSN != InvalidXLogRecPtr &&
+						  MIN_BACKEND_PREFETCH_LSN != GetXLogReplayRecPtr(NULL))) &&
 						!AmPrewarmWorker && /* do not pump prefetch state in prewarm worker */
 						readahead_getpage_pull_timeout_ms > 0;
 
 	if (needs_set != timeout_set)
 	{
-		/* The background writer doens't (shouldn't) read any pages */
-		Assert(!AmBackgroundWriterProcess());
-		/* The checkpointer doens't (shouldn't) read any pages */
-		Assert(!AmCheckpointerProcess());
+		/*
+		 * The background writer/checkpointer doens't (shouldn't) read any pages.
+		 * And definitely they should run on replica.
+		 * The only cae when we can get here is replica promotion.
+		 */
```

Comment:
How would this work? If they don't read pages, how do they get into the `needs_set != timeout_set` condition?

Reply:
It can happen  during replica promotion and added `(MIN_BACKEND_PREFETCH_LSN != InvalidXLogRecPtr && MIN_BACKEND_PREFETCH_LSN != GetXLogReplayRecPtr(NULL))` condition.